# PyCity Schools Analysis

* As a whole, schools with higher budgets, did not yield better test results. By contrast, schools with higher spending per student actually (\$645-675) underperformed compared to schools with smaller budgets (<\$585 per student).

* As a whole, smaller and medium sized schools dramatically out-performed large sized schools on passing math performances (89-91% passing vs 67%).

* As a whole, charter schools out-performed the public district schools across all metrics. However, more analysis will be required to glean if the effect is due to school practices or the fact that charter schools tend to serve smaller student populations per school. 
---

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [9]:
# Dependencies and Setup
import pandas as pd
import numpy as np
import os
# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas Data Frames
school_data_df = pd.read_csv(school_data_to_load)
student_data_df = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset
school_data_complete = pd.merge(student_data_df, school_data_df, how="left", on=["school_name", "school_name"]) 
school_data_complete_df = school_data_complete

In [10]:
student_data_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [11]:
school_data_df.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [12]:
school_data_complete_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [13]:
#school_data_df = pd.read_csv(school_data_to_load)
#student_data_df = pd.read_csv(student_data_to_load)
#school_data_complete_df


#create array of unique school names
School_Names = school_data_df['school_name'].unique()
#gives the length of unique school names to give us how many schools
school_count = len(School_Names)
#district student count
dist_student_count = school_data_df['size'].sum()
#student count from student file (to verify with district student count)
total_student_rec = student_data_df ['student_name'].count()
#total budget
total_budget = school_data_df['budget'].sum()
#calculations for number and % passing reading
num_passing_reading = student_data_df.loc[student_data_df['reading_score'] >= 70]['reading_score'].count()
perc_pass_reading = num_passing_reading/total_student_rec
#calculations for number and % passing math
num_passing_math = student_data_df.loc[student_data_df['math_score'] >= 70]['math_score'].count()
perc_pass_math = num_passing_math/total_student_rec
#average math score calculation
avg_math_score = student_data_df['math_score'].mean()
#average reading score calculation
avg_reading_score = student_data_df['reading_score'].mean()
#Overall Passing Rate Calculations
overall_pass = student_data_df[(student_data_df['math_score'] >= 70) & (student_data_df['reading_score'] >= 70)]['student_name'].count()/total_student_rec

District_Summary = pd.DataFrame({
    
    "Total Schools": [school_count],
    "Total Students": [dist_student_count],
    "Total Budget": [total_budget],
    "Average Math Score": [avg_math_score],
    "Average Reading Score": [avg_reading_score],
    "% Passing Math": [perc_pass_math],
    "% Passing Reading":[perc_pass_reading],
    "Overall Passing Rate": [overall_pass]
})    

#store as different df to change order
dist_sum = District_Summary[["Total Schools", 
                             "Total Students", 
                             "Total Budget", 
                             "Average Math Score", 
                             "Average Reading Score", 
                             '% Passing Math',
                             '% Passing Reading', 
                             'Overall Passing Rate']]
    
#format cells
District_Summary.style.format({"Total Budget": "${:,.2f}",
                       "Average Math Score": "{:.2f}", 
                       "Average Reading Score": "{:.2f}", 
                       "% Passing Math": "{:.2%}", 
                       "% Passing Reading": "{:.2%}", 
                       "Overall Passing Rate": "{:.2%}"})


## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)
  
* Create a dataframe to hold the above results

## Top Performing Schools (By Passing Rate)

* Sort and display the top five schools in overall passing rate

In [14]:
#groups by school
by_school = school_data_complete_df.set_index('school_name').groupby(['school_name'])

#school types
sch_types = school_data_df.set_index('school_name')['type']

# total students by school
stu_per_sch = by_school['Student ID'].count()

# school budget
sch_budget = school_data_df.set_index('school_name')['budget']

#per student budget
stu_budget = school_data_df.set_index('school_name')['budget']/school_data_df.set_index('school_name')['size']

#avg scores by school
avg_math = by_school['math_score'].mean()
avg_read = by_school['reading_score'].mean()

# % passing scores
pass_math = school_data_complete_df[school_data_complete_df['math_score'] >= 70].groupby('school_name')['Student ID'].count()/stu_per_sch 
pass_read = school_data_complete_df[school_data_complete_df['reading_score'] >= 70].groupby('school_name')['Student ID'].count()/stu_per_sch 
overall = school_data_complete_df[(school_data_complete_df['reading_score'] >= 70) & (school_data_complete_df['math_score'] >= 70)].groupby('school_name')['Student ID'].count()/stu_per_sch 

school_summary = pd.DataFrame({
    "School Type": sch_types,
    "Total Students": stu_per_sch,
    "Per Student Budget": stu_budget,
    "Total School Budget": sch_budget,
    "Average Math Score": avg_math,
    "Average Reading Score": avg_read,
    '% Passing Math': pass_math,
    '% Passing Reading': pass_read,
    "Overall Passing Rate": overall
})

school_summary = school_summary[['School Type', 
                          'Total Students', 
                          'Total School Budget', 
                          'Per Student Budget', 
                          'Average Math Score', 
                          'Average Reading Score',
                          '% Passing Math',
                          '% Passing Reading',
                          'Overall Passing Rate']]

#formatting
school_summary.style.format({'Total Students': '{:,}', 
                          "Total School Budget": "${:,}", 
                          "Per Student Budget": "${:.0f}",
                          'Average Math Score': "{:.1f}", 
                          'Average Reading Score': "{:.1f}", 
                          "% Passing Math": "{:.1%}", 
                          "% Passing Reading": "{:.1%}", 
                          "Overall Passing Rate": "{:.1%}"})
school_summary.head()

School Type  Total Students  Total School Budget  \
Bailey High School      District            4976              3124928   
Cabrera High School      Charter            1858              1081356   
Figueroa High School    District            2949              1884411   
Ford High School        District            2739              1763916   
Griffin High School      Charter            1468               917500   

                      Per Student Budget  Average Math Score  \
Bailey High School                 628.0           77.048432   
Cabrera High School                582.0           83.061895   
Figueroa High School               639.0           76.711767   
Ford High School                   644.0           77.102592   
Griffin High School                625.0           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963        0.666801   
Cabrera High School               83.975780        0.941335   
Figueroa High School              81.158020        0.659885   
Ford High School                  80.746258        0.683096   
Griffin High School               83.816757        0.933924   

                      % Passing Reading  Overall Passing Rate  
Bailey High School             0.819333              0.546423  
Cabrera High School            0.970398              0.913348  
Figueroa High School           0.807392              0.532045  
Ford High School               0.792990              0.542899  
Griffin High School            0.971390              0.905995

In [15]:
top_5 = school_summary.sort_values("Overall Passing Rate", ascending = False)
top_5.head()

School Type  Total Students  Total School Budget  \
Cabrera High School     Charter            1858              1081356   
Thomas High School      Charter            1635              1043130   
Griffin High School     Charter            1468               917500   
Wilson High School      Charter            2283              1319574   
Pena High School        Charter             962               585858   

                     Per Student Budget  Average Math Score  \
Cabrera High School               582.0           83.061895   
Thomas High School                638.0           83.418349   
Griffin High School               625.0           83.351499   
Wilson High School                578.0           83.274201   
Pena High School                  609.0           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
Cabrera High School              83.975780        0.941335           0.970398   
Thomas High School               83.848930        0.932722           0.973089   
Griffin High School              83.816757        0.933924           0.971390   
Wilson High School               83.989488        0.938677           0.965396   
Pena High School                 84.044699        0.945946           0.959459   

                     Overall Passing Rate  
Cabrera High School              0.913348  
Thomas High School               0.909480  
Griffin High School              0.905995  
Wilson High School               0.905826  
Pena High School                 0.905405

## Bottom Performing Schools (By Passing Rate)

* Sort and display the five worst-performing schools

In [16]:
bottom_5 = top_5.tail()
bottom_5 = bottom_5.sort_values('Overall Passing Rate')
bottom_5.style.format({'Total Students': '{:,}', 
                       "Total School Budget": "${:,}", 
                       "Per Student Budget": "${:.0f}", 
                       "% Passing Math": "{:.1%}", 
                       "% Passing Reading": "{:.1%}", 
                       "Overall Passing Rate": "{:.1%}"})

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [17]:
#creates grade level average math scores for each school 
ninthgr_math = student_data_df.loc[student_data_df['grade'] == '9th'].groupby('school_name')["math_score"].mean()
tenthgr_math = student_data_df.loc[student_data_df['grade'] == '10th'].groupby('school_name')["math_score"].mean()
eleventhgr_math = student_data_df.loc[student_data_df['grade'] == '11th'].groupby('school_name')["math_score"].mean()
twelfthgr_math = student_data_df.loc[student_data_df['grade'] == '12th'].groupby('school_name')["math_score"].mean()

math_scores = pd.DataFrame({
        "9th": ninthgr_math,
        "10th": tenthgr_math,
        "11th": eleventhgr_math,
        "12th": twelfthgr_math
})
math_scores = math_scores[['9th', '10th', '11th', '12th']]
math_scores.index.name = "school_name"

#show and format
math_scores.style.format({'9th': '{:.2f}', 
                          "10th": '{:.2f}', 
                          "11th": "{:.2f}", 
                          "12th": "{:.2f}"})

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [18]:
#creates grade level average reading scores for each school
ninthgr_reading = student_data_df.loc[student_data_df['grade'] == '9th'].groupby('school_name')["reading_score"].mean()
tenthgr_reading = student_data_df.loc[student_data_df['grade'] == '10th'].groupby('school_name')["reading_score"].mean()
eleventhgr_reading = student_data_df.loc[student_data_df['grade'] == '11th'].groupby('school_name')["reading_score"].mean()
twelfthgr_reading = student_data_df.loc[student_data_df['grade'] == '12th'].groupby('school_name')["reading_score"].mean()

#merges the reading score averages by school and grade together
reading_scores = pd.DataFrame({
        "9th": ninthgr_reading,
        "10th": tenthgr_reading,
        "11th": eleventhgr_reading,
        "12th": twelfthgr_reading
})
reading_scores = reading_scores[['9th', '10th', '11th', '12th']]
reading_scores.index.name = "School"

#format
reading_scores.style.format({'9th': '{:.2f}', 
                             "10th": '{:.2f}', 
                             "11th": "{:.2f}", 
                             "12th": "{:.2f}"})

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [19]:
# Sample bins. Feel free to create your own bins.
spending_bins = [0, 584.999, 614.999, 644.999, 99999]
group_name = ["<$585", "$585-614", "$615-644", ">$644"]

In [20]:
school_data_complete_df['spending_bins'] = pd.cut(school_data_complete_df['budget']/school_data_complete_df['size'], spending_bins, labels = group_name)

#group by spending
by_spending = school_data_complete_df.groupby('spending_bins')

#calculations
avg_math = by_spending['math_score'].mean()
avg_read = by_spending['reading_score'].mean()
pass_math = school_data_complete_df[school_data_complete_df['math_score'] >= 70].groupby('spending_bins')['Student ID'].count()/by_spending['Student ID'].count()
pass_read = school_data_complete_df[school_data_complete_df['reading_score'] >= 70].groupby('spending_bins')['Student ID'].count()/by_spending['Student ID'].count()
overall = school_data_complete_df[(school_data_complete_df['reading_score'] >= 70) & (school_data_complete_df['math_score'] >= 70)].groupby('spending_bins')['Student ID'].count()/by_spending['Student ID'].count()

# df build            
scores_by_spend = pd.DataFrame({
    "Average Math Score": avg_math,
    "Average Reading Score": avg_read,
    '% Passing Math': pass_math,
    '% Passing Reading': pass_read,
    "Overall Passing Rate": overall
            
})

#reorder columns
scores_by_spend = scores_by_spend[[
    "Average Math Score",
    "Average Reading Score",
    '% Passing Math',
    '% Passing Reading',
    "Overall Passing Rate"
]]

scores_by_spend.index.name = "Per Student Budget"
scores_by_spend = scores_by_spend.reindex(group_name)

#formating
scores_by_spend.style.format({'Average Math Score': '{:.2f}', 
                              'Average Reading Score': '{:.2f}', 
                              '% Passing Math': '{:.2%}', 
                              '% Passing Reading':'{:.2%}', 
                              'Overall Passing Rate': '{:.2%}'})


## Scores by School Size

* Perform the same operations as above, based on school size.

In [21]:
# Sample bins. Feel free to create your own bins.
size_bins = [0, 999, 1999, 99999]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [22]:
# Sample bins. Feel free to create your own bins.
size_bins = [0, 999, 1999, 99999]
group_name = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

school_data_complete_df['size_bins'] = pd.cut(school_data_complete_df['size'], size_bins, labels = group_names)

#group by spending
by_size = school_data_complete_df.groupby('size_bins')

#calculations 
avg_math = by_size['math_score'].mean()
avg_read = by_size['math_score'].mean()
pass_math = school_data_complete_df[school_data_complete_df['math_score'] >= 70].groupby('size_bins')['Student ID'].count()/by_size['Student ID'].count()
pass_read = school_data_complete_df[school_data_complete_df['reading_score'] >= 70].groupby('size_bins')['Student ID'].count()/by_size['Student ID'].count()
overall = school_data_complete_df[(school_data_complete_df['reading_score'] >= 70) & (school_data_complete_df['math_score'] >= 70)].groupby('size_bins')['Student ID'].count()/by_size['Student ID'].count()

# df build            
scores_by_size = pd.DataFrame({
    "Average Math Score": avg_math,
    "Average Reading Score": avg_read,
    '% Passing Math': pass_math,
    '% Passing Reading': pass_read,
    "Overall Passing Rate": overall
            
})

#reorder columns
scores_by_size = scores_by_size[[
    "Average Math Score",
    "Average Reading Score",
    '% Passing Math',
    '% Passing Reading',
    "Overall Passing Rate"
]]

scores_by_size.index.name = "Total Students"
scores_by_size = scores_by_size.reindex(group_name)
# scores_by_size.head() 
#formating
scores_by_size.style.format({'Average Math Score': '{:.2f}', 
                              'Average Reading Score': '{:.2f}', 
                              '% Passing Math': '{:.2%}', 
                              '% Passing Reading':'{:.2%}', 
                              'Overall Passing Rate': '{:.2%}'})
            

## Scores by School Type

* Perform the same operations as above, based on school type.

In [23]:
# group by type of school
by_type = school_data_complete_df.groupby("type")

#calculations 
avg_math = by_type['math_score'].mean()
avg_read = by_type['math_score'].mean()
pass_math = school_data_complete_df[school_data_complete_df['math_score'] >= 70].groupby('type')['Student ID'].count()/by_type['Student ID'].count()
pass_read = school_data_complete_df[school_data_complete_df['reading_score'] >= 70].groupby('type')['Student ID'].count()/by_type['Student ID'].count()
overall = school_data_complete_df[(school_data_complete_df['reading_score'] >= 70) & (school_data_complete_df['math_score'] >= 70)].groupby('type')['Student ID'].count()/by_type['Student ID'].count()

# df build            
scores_by_type = pd.DataFrame({
    "Average Math Score": avg_math,
    "Average Reading Score": avg_read,
    '% Passing Math': pass_math,
    '% Passing Reading': pass_read,
    "Overall Passing Rate": overall})
    
#reorder columns
scores_by_type = scores_by_type[[
    "Average Math Score",
    "Average Reading Score",
    '% Passing Math',
    '% Passing Reading',
    "Overall Passing Rate"
]]
scores_by_type.index.name = "Type of School"


#formating
scores_by_type.style.format({'Average Math Score': '{:.1f}', 
                              'Average Reading Score': '{:.1f}', 
                              '% Passing Math': '{:.1%}', 
                              '% Passing Reading':'{:.1%}', 
                              'Overall Passing Rate': '{:.1%}'})

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing Rate  
School Type                                             
Charter              96.586489               95.103660  
District             80.799062               73.673757

Among multiple trends I found in the data there are three I would like to specifically point out. These trends all point towards a student's performance. 

First: 
The larger schools show a distinct drop in performance relative to the medium and small schools. Large = 56.57%, Medium = 90.63% and Small = 90.14%.
Our data sets would indicate that at the 2,000 student mark there's a large influence on performance. Weather your school moves from a "small" or "medium" student size to a 2000+ student set, the school should expect roughly a 33.81% decrease in performance.
    (33.57% + 34.05%)/2 = 33.81% -> average rate of change
    
Second:
The performance differences between the dollars spent on average per individual student. When The spending was binned in to categories a clear trend appeared. When spending per student was betweeen $585 and $614 the overall passing rate for the schools were over 90% Once the dollar amount hit $615 performance dropped to 60% and then 53% when over $644. Concluding that more money doesn't necessarily dictate performance after a certain point. 

Third:
Charter schools were in the top 5 performing via overall passing rate and the bottom five performing were district schools. I would also like to note that all five of the bottom performing schools had a student count of over 2,000 and only one of the top performing charter schools was over the 2,000 student mark. 